docs/spec.md

```diff
@@ -4729,26 +4779,27 @@ Produces a `result` tensor where each element is selected from `on_true` or
 `on_false` tensor based on the value of the corresponding element of `pred`.
 More formally, `result[result_index] = pred_element ? on_true[result_index] :
 on_false[result_index]`, where `pred_element = rank(pred) = 0 ? pred :
-pred[result_index]`.
+pred[result_index]`. For quantized types, performs
+`dequantize_select_quantize(pred, on_true, on_false, type(result))`.
 
 #### Inputs
 
-| Label | Name       | Type                | Constraints |
-|-------|------------|---------------------|-------------|
-| (I1)  | `pred`     | tensor of type `i1` | (C1)        |
-| (I2)  | `on_true`  | tensor              | (C1), (C2)  |
-| (I3)  | `on_false` | tensor              | (C2)        |
+| Label | Name       | Type                                  | Constraints |
+|-------|------------|---------------------------------------|-------------|
+| (I1)  | `pred`     | tensor of type `i1`                   | (C1)        |
+| (I2)  | `on_true`  | tensor or per-tensor quantized tensor | (C1-C2)     |
+| (I3)  | `on_false` | tensor or per-tensor quantized tensor | (C2)        |
 
 #### Outputs
 
-| Name     | Type   | Constraints |
-|----------|--------|-------------|
-| `result` | tensor | (C2)        |
+| Name     | Type                                  | Constraints |
+|----------|---------------------------------------|-------------|
+| `result` | tensor or per-tensor quantized tensor | (C2)        |
 
 #### Constraints
 
 * (C1) `rank(pred) = 0 or shape(pred) = shape(on_true)`.
-* (C2) `type(on_true) = type(on_false) = type(result)`.
+* (C2) `baseline_type(on_true) = baseline_type(on_false) = baseline_type(result)`.
```

Comment:
I think we want to limit rescaling in select as well, thus instead of using `baseline_type(...)`, we should keep this as `element_type(...)`

Reply:
Hi @loganchien
This "extra" general behavior of select is to draw parity with other element wise operations. IMHO, this allows treating all op in the category uniformly. Let me know if is makes sense.

Reply:
If we allow different LHS/RHS/Output Scale/Zp, then we will need a transformation to rescale input scales (LHS/RHS) to the output scale (just like what we handled `tfl.maximum` before the TFLite team added the constraint to `tfl.maximum`):

```
%lhs_rescaled = "stablehlo.quantize_uniform"(%lhs)
    : (tensor<... x !quant.uniform<ui8:f32, lhs_scale:lhs_zp>>)
    -> tensor<... x !qunat.uniform<ui8:f32, out_scale:out_zp>>
%rhs_rescaled = "stablehlo.quantize_uniform"(%rhs)
    : (tensor<... x !quant.uniform<ui8:f32, rhs_scale:rhs_zp>>)
    -> tensor<... x !qunat.uniform<ui8:f32, out_scale:out_zp>>
%out = "stablehlo.select"(%cond, %lhs_rescaled, %rhs_rescaled) : ...
```

So it is the question on whether we want to prefer same scale (and explictly rescale when why are different) or to prefer different scale (and internally rescale to output scale if the backend doesn't have other means).

Reply:
Thanks @loganchien for clearly specificying the trade-offs.

The PR in the current form prefers the second of the two you mentioned. Actually, this is the same approach we used for almost all the element wise operations in this PR, where in, the explicit re-scaling  of operands are avoided; the op semantics will internally take care of it and, as you said, the backend will rescale if no other means are available.

Reply:
Most common implementations I am aware of: (1) use trivial element copy while selection, (2) use `cond * a + (!cond) * b` or `(mask(cond) & a) | (mask(~cond) & b)`, or  (3) SIMD instructions (AArch64 `SEL`, RISCV `VMERGE`, etc) prefers the scales to be the same.

But probably it is fine.  Seeing different scales is not the end of day.  The backend can always apply the transformation I mentioned earlier to ensure all scales are same.

This can be marked as resolved.

Reply:
Thanks for the additional insights.